### DISCUSSION TITLE
exact data send for mcu protocol

### DISCUSSION THREAD

**Discussion by benwadub** [tttapa/Control-Surface#622] [Q&A]
hi, I try to make a script for Ableton live, I got two questions:
1: I use remotely to make the script, that create listener and we can have a reaction from this,
remotely ask me the raw midi data sent by the Ableton , I use the mcu protocol in the library I found that the mute1 is 0x10 if I understand well, how can I know what are the correspondences with this number ( it s hexadecimal or a sysex?) and the number note.
I tried to use the reverse mcu code to extract the data sent but it seem to not work so I d like to find where I can find the exact data sent from the mcu protocol to the controller for it recognize as a MCU message?
I already looked at the mcu document you already gave to the issues

2 if I use the MCU script in Ableton all my leds are working fine on normal track midi or audio but the return track (also called aux tracks doesn't t respond to the MCU note, any idea on what messages are sent from Ableton for theses tracks?
thanks
ps my first question is the most critical one for me!

**Comment by tttapa** [comment]
You're referring to these resources, right?

- https://images.thomann.de/pics/prod/151261_manual.pdf#page=120
- https://tttapa.github.io/Control-Surface-doc/Doxygen/d3/d85/Mackie-Control-Universal-Reverse-Engineering_8ino-example.html
- https://tttapa.github.io/Control-Surface-doc/Doxygen/de/d44/group__MCU__Notes.html#gaf88b8bb23e0990b4cc5b95de4431acc0
- https://tttapa.github.io/Control-Surface-doc/Doxygen/de/d44/group__MCU__Notes.html#var-members

If they don't answer your question, I'm afraid I don't really understand what you're asking.

> (it s hexadecimal or a sysex?)

Neither. A number is a number, 0x10 is the same number as 16, the Arduino cannot distinguish between the two representations.

> but it seem to not work

What doesn't work?

> 2 if I use the MCU script in Ableton all my leds are working fine on normal track midi or audio but the return track (also called aux tracks doesn't t respond to the MCU note, any idea on what messages are sent from Ableton for theses tracks?

None, probably. Unless you're in the MCU “AUX” or “Busses” view perhaps.

You should use a MIDI monitor or one of the monitoring or reverse engineering sketches to find out which messages are sent by the DAW.

**Reply by benwadub** [reply]
Thanks for your answer. I was not really clear sorry 😅.
I d like to know what your lubrary need to receive to understand it as a mcu data.
It s a note on message on note 16 in the midi note it s that? It as to be received in decimal?

**Reply by tttapa** [reply]
> I d like to know what your lubrary need to receive to understand it as a mcu data.

Nothing. There is no way to distinguish between an MCU mute message and an ordinary MIDI note on message.

Interpretation of the messages is determined by the software (the Arduino code you write and the settings in the DAW).

> It s a note on message on note 16 in the midi note it s that?

Yes. But if you have no additional context, you have no way to know whether that message originated from someone pressing a key on a MIDI keyboard/piano or from someone pressing a MUTE button on an MCU.

> It as to be received in decimal?

A number is a number, there is no "decimal" in MIDI communication. Decimal and hexadecimal are just ways to represent numbers for humans, Arduinos and computers represent numbers using binary, and the MIDI messages are sent as a sequence of binary bits, so if anything, you could say they are received in binary. But you as a programmer don't care, it's just a number, you don't see the base until you decide to convert it to a human-readable representation again.

For example, 0b1010, 0xA, 10, "ten", "dix" ... are just different names/representations for the same abstract concept of 
"the number of fingers a typical human has". The meaning of a number doesn't change when you change the base.

**Reply by benwadub** [reply]
Thanks I got it!!
All is working nicely now!!
I m waiting for my rgb led to try the vumeter and the project will be done!! I really d like to send you something to thank you for all your work, maybe I can send you something via PayPal or maybe few piece of hardware like an ili9341 or other thing you need to continue developing.
Let me know how I can thank you for this beast I built! Without your work It could never happened!

**Reply by tttapa** [reply]
Glad to hear it works now, it looks great!

I do appreciate the offer, but I'm afraid I don't have much time right now to continue development or provide support, so I can't really accept it.

**Reply by benwadub** [reply]
I could also send you few euros just to thank you !

**Comment by benwadub** [comment]
![D185D14B-729F-4D30-8890-A3AD80C0C4E1](https://user-images.githubusercontent.com/59668816/141696634-107c6498-c3b3-4fbe-9c75-2564b9cf9aa5.jpeg)
